packages/connector/src/services/core/middleware/balance.ts

```diff
@@ -15,19 +15,20 @@ export function createBalanceMiddleware() {
     }
 
     // Update balances on prepare
-    await services.accounts.adjustBalances({
+    const trxOrError = await services.accounts.transferFunds({
       sourceAccountId: accounts.incoming.accountId,
       destinationAccountId: accounts.outgoing.accountId,
-      sourceAmount: BigInt(amount),
-      callback: async (trx: Transaction) => {
-        await next()
+      sourceAmount: BigInt(amount)
+    })
+
+    await next()
 
-        if (response.fulfill) {
-          await trx.commit()
-        } else {
-          await trx.rollback()
-        }
+    if (!isTransferError(trxOrError)) {
+      if (response.fulfill) {
+        await (trxOrError as Transaction).commit()
```

Comment:
Nit: can these casts be removed? I think the `!isTransferError` narrows the type to `Transaction`